src/Microsoft.VisualStudio.Threading/JoinableTaskDependencyGraph.cs

```diff
@@ -453,7 +453,7 @@ internal static void AddSelfAndDescendentOrJoinedJobs(IJoinableTaskDependent tas
 
                 if (taskOrCollection is JoinableTask thisJoinableTask)
                 {
-                    if (thisJoinableTask.IsFullyCompleted || !joinables.Add(thisJoinableTask))
+                    if (thisJoinableTask.IsFullyCompleted || !joinables.Add(thisJoinableTask) || thisJoinableTask.IsCompleteRequested)
```

Comment:
Is this expression in the optimal order? Or should `|| thisJoinableTask.IsCompleteRequested` be added next to the `IsFullyCompleted` check, and before adding to a collection which sounds heavier than a simple bool check.
`joinables.Add` would have side-effects though, so let's be careful about when the side effect should occur vs when `|| thisJoinableTask.IsCompleteRequested` is evaluated.

Reply:
yeah, we need add thisJoinableTask to the collection, if it is not full completed. But the IsFullyCompleted did an extra NoMessagePumpSynchronizationContext.Apply (it is in the lock, so no extra overhead for the contention.) I guess we can move IsCompleteRequested earlier because it cheapest, so it will have more lines of code, but potentially faster.